File: interviewBit-golang/src/backtracking/problems/allUniquePermutations.go
package backtracking

/*
	Problem : https://www.interviewbit.com/problems/all-unique-permutations/
*/

func solveAllUniquePermutations(A map[int]int, B []int, C *[][]int, D int) {
	if len(B) == D {
		E := make([]int, D)
		copy(E, B)
		*C = append(*C, E)
		return
	}

	for k, v := range A {
		if v == 0 {
			continue
		}
		A[k]--
		B = append(B, k)
		solveAllUniquePermutations(A, B, C, D)
		A[k]++
		B = B[:len(B)-1]
	}
}

// T(n) : O() ; S(n) : O()
func AllUniquePermutations(A []int) [][]int {

	n := len(A)
	res := make([][]int, 0)
	if n == 0 {
		return res
	}

	m := make(map[int]int, 0)
	for _, v := range A {
		m[v]++
	}
	t := make([]int, 0)
	solveAllUniquePermutations(m, t, &res, n)
	return res
}

File: interviewBit-cpp/src/backtracking/all-unique-permutations.cpp
#include <bits/stdc++.h>

using namespace std;

void allUniquePermutationsRecursive(vector<int> &A, int idx, vector<int> &temp, vector<vector<int> > &permutations, unordered_map<int, int> &um) {

    if (idx == A.size()) {
        permutations.push_back(temp);
        return ;
    }

    for (auto &[k, v] : um) {
        if (v == 0) continue;
        um[k]--;
        temp.push_back(k);
        allUniquePermutationsRecursive(A, idx + 1, temp, permutations, um);
        um[k]++;
        temp.pop_back();
    }

    return ;
}

// https://www.interviewbit.com/problems/all-unique-permutations/
// T(n) : O(n*n!) ; S(n) : O(n)
vector<vector<int> > solveAllUniquePermutations(vector<int> &A) {

    int n = A.size();
    vector<vector<int> > allPermutations;
    if (n == 0) return allPermutations;

    unordered_map<int, int> um;
    for (auto &a : A) um[a]++;

    vector<int> temp;
    allUniquePermutationsRecursive(A, 0, temp, allPermutations, um);
    return allPermutations;
}

// Driver Code for testing
int main() {


    return 0;
}
